src/Runner/PoolRunner.php

```diff
@@ -2,103 +2,108 @@
+use Psr\Http\Client\ClientInterface;
+use Psr\Http\Message\RequestFactoryInterface;
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+use function Amp\coroutine;
 
 class PoolRunner
 {
-    /** @var TestWorkflow */
-    private $workflow;
+    private Semaphore $semaphore;
 
-    /** @var RequestFactory */
-    private $requestFactory;
+    private ClientInterface $client;
 
-    /** @var Semaphore */
-    private $semaphore;
-
-    /** @var bool */
-    private $allowSelfSignedCertificate;
-
-    public function __construct(RequestFactory $requestFactory, TestWorkflow $workflow, $concurrency = 10, bool $allowSelfSignedCertificate = false)
+    public function __construct(private RequestFactoryInterface $requestFactory, private ResponseFactoryInterface $responseFactory, private StreamFactoryInterface $streamFactory,  private TestWorkflow $workflow, $concurrency = 10, private bool $allowSelfSignedCertificate = false)
     {
-        $this->requestFactory = $requestFactory;
-        $this->workflow = $workflow;
         $this->semaphore = new LocalSemaphore($concurrency);
-        $this->allowSelfSignedCertificate = $allowSelfSignedCertificate;
+
+        $tlsContext = new ClientTlsContext('');
+
+        if ($allowSelfSignedCertificate) {
+            $tlsContext = $tlsContext->withoutPeerVerification();
+        }
+
+        $connectContext = new ConnectContext('');
+        $connectContext = $connectContext->withTlsContext($tlsContext);
+
+        $builder = new HttpClientBuilder();
+        $builder = $builder->usingPool(new UnlimitedConnectionPool(new DefaultConnectionFactory(null, $connectContext)));
+        $client = $builder->build();
+        $adapter = new PsrAdapter($requestFactory, $responseFactory);
+        $this->client = new PsrHttpClient($client, $adapter);
     }
 
     public function loop(Pool $pool)
     {
-        return \Amp\call(function () use ($pool) {
-            ob_start();
-            $promises = [];
+        ob_start();
+        /** @var Future[] $futures */
+        $futures = [];
 
-            while (!$pool->isEmpty()) {
-                $test = $pool->getTestToRun();
+        while (!$pool->isEmpty()) {
+            $test = $pool->getTestToRun();
 
-                if (null === $test) {
-                    yield \Amp\Promise\first($promises);
+            if (null === $test) {
+                Future\any($futures);
 
-                    continue;
-                }
-
-                $promises[$test->getIdentifier()] = $this->run($test);
-                $promises[$test->getIdentifier()]->onResolve(function () use (&$promises, $test) {
-                    unset($promises[$test->getIdentifier()]);
-                });
+                continue;
             }
 
-            yield $promises;
+            $this->workflow->markTestAsRunning($test);
+
+            $futures[$test->getIdentifier()] = coroutine(function () use($test, &$futures) {
```

Comment:
corouting function does not exist (I just cloned the project and I ran `composer up`)

Reply:
It's been renamed to `launch`